### PR TITLE
[Enhancement] Hide the name of the student when downloading repos for Programming Assessment

### PR DESCRIPTION
**Pull request by fde312**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

This PR enables the possibility to hide the students name when downloading the repo. For the manual assessment for programming exercises we then can assure that the assessor is not biased when downloading the repo.

### Description

- Added boolean check if student name needs to be hidden
- Added alternative folder name when student name needs to be hidden

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Assess a programming exercise
4. Click on `Download repo`and download the repo (test different options)
5. Downloaded repo should not contain information about the student

### Test Coverage

Nothing changed

### Screenshots

image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 1]
```diff
@@ -657,23 +657,30 @@ public void deleteLocalRepository(URL repoUrl, String targetPath) {
...
-        // take the last component
-        String zipRepoName = courseShortName + "-" + repositoryUrlComponents[repositoryUrlComponents.length - 1] + ".zip";
+        String zipRepoName;
+        if (hideStudentName) {
+            zipRepoName = courseShortName + "-" + repositoryUrlComponents[repositoryUrlComponents.length - 2].toLowerCase() + "-student-submission" + ".zip";
+        }
+        else {
+            // take the last component
+            zipRepoName = courseShortName + "-" + repositoryUrlComponents[repositoryUrlComponents.length - 1] + ".zip";
+        }
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:683]: Can you assign `repositoryUrlComponents[n]` to a variable with a meaningful name?
Would help if we need to touch this again in a week or so :D

File: src/main/webapp/app/exercises/programming/assess/repo-export/programming-assessment-repo-export-dialog.component.ts [hunk 2]
```diff
@@ -75,16 +76,18 @@ export class ProgrammingAssessmentRepoExportDialogComponent implements OnInit {
         if (this.participationIdList) {
             // We anonymize the assessment process ("double-blind").
             this.repositoryExportOptions.addParticipantName = false;
-            this.repoExportService.exportReposByParticipations(exerciseId, this.participationIdList, this.repositoryExportOptions).subscribe(this.handleExportRepoResponse, () => {
-                this.exportInProgress = false;
-            });
+            this.repoExportService
+                .exportReposByParticipations(exerciseId, this.participationIdList, this.repositoryExportOptions, this.hideStudentName)
```
Comment (jpbernius) [inline src/main/webapp/app/exercises/programming/assess/repo-export/programming-assessment-repo-export-dialog.component.ts:80]: What are the export options? Hiding a student name sounds like an export option.

**Review by jpbernius — CHANGES_REQUESTED** [review]
Two requests.
In general: good changes. Can you consider to add the hide name option as part of the repository export options?

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 3]
```diff
@@ -657,23 +657,30 @@ public void deleteLocalRepository(URL repoUrl, String targetPath) {
...
-    public Path zipRepository(Repository repo, String targetPath) throws IOException {
+    public Path zipRepository(Repository repo, String targetPath, boolean hideStudentName) throws IOException {
         String[] repositoryUrlComponents = repo.getParticipation().getRepositoryUrl().split(File.separator);
         ProgrammingExercise exercise = repo.getParticipation().getProgrammingExercise();
         String courseShortName = exercise.getCourseViaExerciseGroupOrCourseMember().getShortName().replaceAll("\\s", "");
-        // take the last component
-        String zipRepoName = courseShortName + "-" + repositoryUrlComponents[repositoryUrlComponents.length - 1] + ".zip";
+        String zipRepoName;
+        if (hideStudentName) {
+            zipRepoName = courseShortName + "-" + repositoryUrlComponents[repositoryUrlComponents.length - 2].toLowerCase() + "-student-submission" + ".zip";
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:678]: not sure if I understand what's going on here. Can you document this a bit better?

**Review by tiffca — APPROVED** [review]
Tested successfully on test server 3, confirmed the student's name is not provided. Instead I see a random integer string

**Comment by derLalla** [comment]
> Tested successfully on test server 3, confirmed the student's name is not provided. Instead I see a random integer string
> 
> 

you need to unzip it and look inside if no student name is visible

**Review by derLalla — CHANGES_REQUESTED** [review]
no matter what I am selecting, always the student name is exported. 

image

This is the correct setting to not get the student names, right?

**Review by derLalla — APPROVED** [review]
I've accidentally tested the wrong part, everything is working fine as intended! :)

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java (hunk 3 above)
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java]: Done in `d4a225b`

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java (hunk 1 above)
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java]: I added detailed comments what we are using in `d4a225b`

File: src/main/webapp/app/exercises/programming/assess/repo-export/programming-assessment-repo-export-dialog.component.ts (hunk 2 above)
Comment (fde312) [reply src/main/webapp/app/exercises/programming/assess/repo-export/programming-assessment-repo-export-dialog.component.ts]: good catch! I just refactored it to use the export options

**Review by tiffca — APPROVED** [review]
Tested successfully again on test server 3, tried all of the options and the student name is not shown

**Review by FrankeLukas — APPROVED** [review]
Code changes lgtm but I can't download repositories on the test server (no idea why; it happens on other branches too), so :+0.5:

**Review by derLalla — APPROVED** [review]
re-approve. Everything works as expected! :)

**Review by sjagla27 — APPROVED** [review]
Could not find any stundent names in downloaded repositories.

**Review by schultek — APPROVED** [review]
The student id is not visible in the file or folder names directly and the code looks good. Therefore I approve.

However as a consideration for a future PR: The tutor could still figure out the student id when inspecting the hidden .git directory since the full origin url (e.g. of bitbucket) still exists in the git config file.

**Review by FuchsDominik — APPROVED** [review]
Couldnt find any student detection. (Exept maybe Kilians remark) So i think thats fine.
Also team exercises do not show any team name, however maybe it would be possible to rename it to ...team-submission... if its a team exercise? If thats possible
image

**Review by balazs-czopf — APPROVED** [review]
Looks good to me, nice work 🚀

**Pull request merged**